### PR TITLE
Markov Chain accepts matrix PR#748

### PR DESCRIPTION
**Pull request by kris-singh**

### CODE REVIEW

**Comment by kris-singh** [comment]
All nose-test cases are passing for me. I don't understand the errors. Can anyone help

**Comment by ankurankan** [comment]
@kris-singh You haven't imported numpy in the test file. It shouldn't pass on your local machine also.

**Comment by khalibartan** [comment]
@kris-singh Can you remove the .pgmx file from the commit? It is not recommended to use `git add .` while adding the files for commit.I guess you might have used it.

**Comment by kris-singh** [comment]
@khalibartan yes i will do it right now

File: pgmpy/models/MarkovChain.py [hunk 1]
```diff
@@ -186,11 +186,23 @@ def add_transition_model(self, variable, transition_model):
...
         >>> model.add_variable('grade', 3)
         >>> grade_tm = {0: {0: 0.1, 1: 0.5, 2: 0.4}, 1: {0: 0.2, 1: 0.2, 2: 0.6 }, 2: {0: 0.7, 1: 0.15, 2: 0.15}}
+        >>> grade_tm_matrix = np.array([[0.1, 0.5, 0.4], [0.2, 0.2, 0.6], [0.7, 0.15, 0.15]])
         >>> model.add_transition_model('grade', grade_tm)
+        >>> model.add_transition_model('grade', grade_tm_matrix)
         """
         # check if the transition model is valid
-        if not isinstance(transition_model, dict):
-            raise ValueError('Transition model must be a dict.')
+        if not isinstance(transition_model, dict) and not isinstance(transition_model, np.ndarray):
```
Comment (ankurankan) [inline pgmpy/models/MarkovChain.py]: @kris-singh Add a statement to convert `transition_model` to a numpy array in the case when the user passes a 2-D list.

File: pgmpy/models/MarkovChain.py [hunk 2]
```diff
@@ -186,11 +186,23 @@ def add_transition_model(self, variable, transition_model):
...
+            raise ValueError('Transition model must be a dict or numpy array')
+        if isinstance(transition_model, np.ndarray) and len(transition_model.shape) != 2:
+            raise ValueError('Must be 2d array')
+        if isinstance(transition_model, np.ndarray) and transition_model.shape[0] != transition_model.shape[1]:
+            raise ValueError('Dimension mismatch{d1}!={d2}'.format(d1=transition_model.shape[0], d2=transition_model[1]))
```
Comment (ankurankan) [inline pgmpy/models/MarkovChain.py]: @kris-singh Much cleaner if-else would be:
```python
if not dict:
    if not array:
         raise
   elif shape:
         raise
   elif shape = sh:
        raise
```

File: pgmpy/tests/test_models/test_MarkovChain.py [hunk 3]
```diff
@@ -149,15 +152,17 @@ def test_add_variables_from(self, add_var):
...
         self.assertRaises(ValueError, model.add_transition_model, 'var', 123)
+        self.assertRaises(ValueError, model.add_transition_model, 'var', grade_tm_matrix_bad)
 
     def test_add_transition_model_bad_states(self):
         model = MC(['var'], [2])
         # transition for state=1 not defined
         transition_model = {0: {0: 0.1, 1: 0.9}}
         self.assertRaises(ValueError, model.add_transition_model, 'var', transition_model)
-
+
```
Comment (ankurankan) [inline pgmpy/tests/test_models/test_MarkovChain.py]: @kris-singh Blank line has spaces.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by ankurankan** [comment]
@khalibartan @yashu-seth Can you please review this also ? @kris-singh Can you please merge all your commits into a single commit ?

**Comment by ankurankan** [comment]
@kris-singh Since you are working with Markov Chains, I think you might also want to have a look at #789.

File: pgmpy/models/MarkovChain.py [hunk 4]
```diff
@@ -186,11 +186,23 @@ def add_transition_model(self, variable, transition_model):
...
         >>> model.add_transition_model('grade', grade_tm)
+        >>> model.add_transition_model('grade', grade_tm_matrix)
         """
         # check if the transition model is valid
-        if not isinstance(transition_model, dict):
-            raise ValueError('Transition model must be a dict.')
+        if not isinstance(transition_model, dict) and not isinstance(transition_model, np.ndarray):
+            raise ValueError('Transition model must be a dict or numpy array')
+        if isinstance(transition_model, np.ndarray) and len(transition_model.shape) != 2:
+            raise ValueError('Must be 2d array')
```
Comment (khalibartan) [inline pgmpy/models/MarkovChain.py]: @kris-singh IMO this error is a bit ambiguous. I think something like `Transition model must be a 2d array like object` should be better. The current error doesn't link it to transition model. Please change that.

Comment (khalibartan) [inline pgmpy/models/MarkovChain.py]: @kris-singh IMO this error is a bit ambiguous. I think something like Transition model must be a 2d array like object should be better. The current error doesn't link it to transition model. Please change that.

File: pgmpy/models/MarkovChain.py [hunk 5]
```diff
@@ -186,11 +186,23 @@ def add_transition_model(self, variable, transition_model):
...
+        if isinstance(transition_model, np.ndarray) and transition_model.shape[0] != transition_model.shape[1]:
+            raise ValueError('Dimension mismatch{d1}!={d2}'.format(d1=transition_model.shape[0], d2=transition_model[1]))
+
+        # convert the matrix to dict
+        if isinstance(transition_model, np.ndarray):
+            size = transition_model.shape[0]
+            transition_model = dict((i,dict((j, float(transition_model[i][j]))\
```
Comment (yashu-seth) [inline pgmpy/models/MarkovChain.py]: @kris-singh PEP8
missing whitespace after ','
the backslash is redundant between brackets

File: pgmpy/tests/test_models/test_MarkovChain.py [hunk 6]
```diff
@@ -181,6 +186,21 @@ def test_add_transition_model_success(self):
         model.add_transition_model('var', transition_model)
         self.assertDictEqual(model.transition_models['var'], transition_model)
 
+    def test_transition_model_bad_matrix_dimension(self):
+        model = MC(['var'], [2])
+        transition_model = np.array([0.3, 0.7])
+        #check for square dimension of the matrix
```
Comment (yashu-seth) [inline pgmpy/tests/test_models/test_MarkovChain.py]: @kris-singh PEP8 block comment should start with '# '

**Review by yashu-seth — CHANGES_REQUESTED** [review]

**Comment by kris-singh** [comment]
Sorry i made a very silly mistake using squash and had to revert back and then commit :) thats why all the coverall comments.

File: pgmpy/models/MarkovChain.py [hunk 7]
```diff
@@ -186,11 +186,28 @@ def add_transition_model(self, variable, transition_model):
...
+            if not isinstance(transition_model, np.ndarray):
+                raise ValueError('Transition model must be a dict or numpy array')
+            elif len(transition_model.shape) != 2:
+                raise ValueError('Transition model must be 2d array.given {t}'.format(t=transition_model.shape))
+            elif transition_model.shape[0] != transition_model.shape[1]:
+                raise ValueError('Dimension mismatch {d1}!={d2}'.format(d1=transition_model.shape[0],
+                                 d2=transition_model.shape[1]))
+
+        # convert the matrix to dict
+        if isinstance(transition_model, np.ndarray):
```
Comment (ankurankan) [inline pgmpy/models/MarkovChain.py]: @kris-singh Can't we put this condition also in the above checks?

File: pgmpy/models/MarkovChain.py [hunk 8]
```diff
@@ -177,7 +177,7 @@ def add_transition_model(self, variable, transition_model):
         -----------
         variable: any hashable python object
             must be an existing variable of the model.
-        transition_model: dict
+        transition_model: dict or ndarray
```
Comment (ankurankan) [inline pgmpy/models/MarkovChain.py]: @kris-singh  Add a more detailed description for the parameter explaining what kind of matrix or dict is expected.

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/models/MarkovChain.py (hunk 8 above)
Comment (kris-singh) [reply pgmpy/models/MarkovChain.py]: I think this line only tells the type of object that is expected which in this case is dict,2d-array.
so i think i could replace this with
transition_model: dict or 2d-array
The next line provides the description which i thik is apt

Comment (ankurankan) [reply pgmpy/models/MarkovChain.py]: @kris-singh I think this line should mention about 2-D array. And the description should mention the structure of the dict or array that we are expecting. Check other parts of the code for examples.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by kris-singh** [comment]
@ankurankan @khalibartan can you review this guys.

**Review by ankurankan — APPROVED** [review]

**Pull request merged**